src/common/hwio_a3ides_2209_02.cpp

```diff
@@ -550,80 +564,89 @@ void hwio_arduino_error(int err, uint32_t pin32) {
-            return;
-        case MARLIN_PIN(FAN1):
-            //hwio_fan_set_pwm(_FAN1, ulVal?255:0);
-            //_hwio_pwm_analogWrite_set_val(HWIO_PWM_FAN1, ulVal ? _pwm_analogWrite_max[HWIO_PWM_FAN1] : 0);
+/**
+ * @brief Write digital pin to be used from Marlin
+ *
+ * Use MARLIN_PIN macro when handling special cases (virtual pins)
+ * @example @code
+ * case MARLIN_PIN(FAN):
+ * @endcode
+ *
+ */
+void digitalWrite(uint32_t marlinPin, uint32_t ulVal) {
+#if _DEBUG
+    if (!HAL_GPIO_Initialized) {
+        hwio_arduino_error(HWIO_ERR_UNINI_DIG_WR, marlinPin); //error: uninitialized digital write
+        return;
+    }
+#endif //_DEBUG
+    switch (marlinPin) {
+    case MARLIN_PIN(BED_HEAT):
+        _hwio_pwm_analogWrite_set_val(HWIO_PWM_HEATER_BED, ulVal ? _pwm_analogWrite_max[HWIO_PWM_HEATER_BED] : 0);
+        return;
+    case MARLIN_PIN(HEAT0):
+        _hwio_pwm_analogWrite_set_val(HWIO_PWM_HEATER_0, ulVal ? _pwm_analogWrite_max[HWIO_PWM_HEATER_0] : 0);
+        return;
+    case MARLIN_PIN(FAN1):
 #ifdef NEW_FANCTL
-            if (hwio_fan_control_enabled)
-                fanctl_set_pwm(1, ulVal ? (100 * 50 / 255) : 0);
+        if (hwio_fan_control_enabled)
+            fanctl_set_pwm(1, ulVal ? (100 * 50 / 255) : 0);
 #else  //NEW_FANCTL
-            _hwio_pwm_analogWrite_set_val(HWIO_PWM_FAN1, ulVal ? 100 : 0);
+        _hwio_pwm_analogWrite_set_val(HWIO_PWM_FAN1, ulVal ? 100 : 0);
 #endif //NEW_FANCTL
-            return;
-        case MARLIN_PIN(FAN):
+        return;
+    case MARLIN_PIN(FAN):
 #ifdef NEW_FANCTL
-            if (hwio_fan_control_enabled)
-                fanctl_set_pwm(0, ulVal ? 50 : 0);
+        if (hwio_fan_control_enabled)
+            fanctl_set_pwm(0, ulVal ? 50 : 0);
 #else  //NEW_FANCTL
-            _hwio_pwm_analogWrite_set_val(HWIO_PWM_FAN, ulVal ? _pwm_analogWrite_max[HWIO_PWM_FAN] : 0);
+        _hwio_pwm_analogWrite_set_val(HWIO_PWM_FAN, ulVal ? _pwm_analogWrite_max[HWIO_PWM_FAN] : 0);
 #endif //NEW_FANCTL
-            return;
-        case MARLIN_PIN(X_DIR):
-        case MARLIN_PIN(X_STEP):
-        case MARLIN_PIN(Z_ENA):
-        case MARLIN_PIN(X_ENA):
-        case MARLIN_PIN(Z_STEP):
-        case MARLIN_PIN(E0_DIR):
-        case MARLIN_PIN(E0_STEP):
-        case MARLIN_PIN(E0_ENA):
-        case MARLIN_PIN(Y_DIR):
-        case MARLIN_PIN(Y_STEP):
-#if (MARLIN_PIN(X_ENA) != MARLIN_PIN(Y_ENA))
-        case MARLIN_PIN(Y_ENA):
-#endif
-        case MARLIN_PIN(Z_DIR):
-            gpio_set(ulPin, ulVal ? 1 : 0);
-            return;
-        default:
-            hwio_arduino_error(HWIO_ERR_UNDEF_DIG_WR, ulPin); //error: undefined pin digital write
+        return;
+    default:
+#if _DEBUG
+        if (buddy::hw::isOutputPin(marlinPin)) {
+#endif //_DEBUG
+            gpio_set(marlinPin, ulVal ? 1 : 0);
+#if _DEBUG
+        } else {
+            hwio_arduino_error(HWIO_ERR_UNDEF_DIG_WR, marlinPin); //error: undefined pin digital write
         }
```

Comment:
/suggestion/ what about a small change to make it more readable?

```
#if _DEBUG
    if (!isOutputPin)
        hwio_arduino_error(...);
#endif
gpio_set(...)
```

Reply:
Fixed as suggested.